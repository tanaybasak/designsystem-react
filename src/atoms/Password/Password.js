import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import Tooltip from '../Tooltip/Tooltip';
import prefix from '../../settings';

const Password = ({ className, disabled, ...restProps }) => {
  const [value, setValue] = useState(restProps.value || '');
  const [isIconVisible, toggleIconVisibility] = useState(false);
  const [typechange, setType] = useState('password');
  const classnames = `${prefix}-form-control ${className}`.trim();
  const inputRef = useRef(null);

  const passwordIcon = (
    <button
      type="button"
      disabled={disabled}
      className="hcl-password-visibility"
      aria-label="Show password"
      onClick={() => {
        toggleIconVisibility(!isIconVisible);
        inputRef && inputRef.current ? inputRef.current.focus() : null;
      }}
    >
      {disabled ? (
        <>
          <svg
            className="hcl-icon-visibility-off"
            width="16"
            height="16"
            viewBox="0 0 16 16"
            style={isIconVisible ? { display: 'none' } : { display: 'initial' }}
          >
            <path d="M11.846 3.45L15.293.007 16 .714l-3.284 3.281c1.261.902 2.377 2.212 3.347 3.93C14.02 11.642 11.333 13.5 8 13.5c-1.392 0-2.667-.324-3.822-.973L.703 16l-.706-.708 3.323-3.32C2.071 11.042.976 9.694.035 7.924 2.012 4.308 4.667 2.5 8 2.5c1.395 0 2.677.317 3.846.95zm-6.928 8.338c.944.477 1.97.712 3.082.712 2.795 0 5.076-1.483 6.907-4.568-.866-1.417-1.833-2.486-2.91-3.219l-1.55 1.55a3 3 0 0 1-4.185 4.182l-1.344 1.343zm-.882-.533l1.518-1.517A3 3 0 0 1 9.74 5.556l1.364-1.363A7.02 7.02 0 0 0 8 3.5c-2.798 0-5.047 1.439-6.819 4.432.842 1.465 1.792 2.568 2.855 3.323zm2.948-1.532a2 2 0 0 0 2.74-2.738l-2.74 2.738zm-.707-.707l2.74-2.738a2 2 0 0 0-2.74 2.738z" />
          </svg>
          <svg
            className="hcl-icon-visibility-on"
            aria-label="Hide password"
            width="16"
            height="16"
            viewBox="0 0 16 11"
            style={
              !isIconVisible ? { display: 'none' } : { display: 'initial' }
            }
          >
            <path d="M8 7.5c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 1c-1.7 0-3-1.3-3-3s1.3-3 3-3 3 1.3 3 3-1.3 3-3 3z" />
            <path d="M8 10c2.8 0 5.1-1.5 6.9-4.6C13.1 2.5 10.8 1 8 1 5.2 1 3 2.4 1.2 5.4 2.9 8.6 5.2 10 8 10zM8 0c3.3 0 6 1.8 8.1 5.4C14 9.2 11.3 11 8 11S2 9.2 0 5.5C2 1.9 4.6 0 8 0z" />
          </svg>
        </>
      ) : (
        <>
          <Tooltip content="Show Password" direction="top" type="icon">
            <svg
              className="hcl-icon-visibility-off"
              width="16"
              height="16"
              viewBox="0 0 16 16"
              style={
                isIconVisible ? { display: 'none' } : { display: 'initial' }
              }
            >
              <path d="M11.846 3.45L15.293.007 16 .714l-3.284 3.281c1.261.902 2.377 2.212 3.347 3.93C14.02 11.642 11.333 13.5 8 13.5c-1.392 0-2.667-.324-3.822-.973L.703 16l-.706-.708 3.323-3.32C2.071 11.042.976 9.694.035 7.924 2.012 4.308 4.667 2.5 8 2.5c1.395 0 2.677.317 3.846.95zm-6.928 8.338c.944.477 1.97.712 3.082.712 2.795 0 5.076-1.483 6.907-4.568-.866-1.417-1.833-2.486-2.91-3.219l-1.55 1.55a3 3 0 0 1-4.185 4.182l-1.344 1.343zm-.882-.533l1.518-1.517A3 3 0 0 1 9.74 5.556l1.364-1.363A7.02 7.02 0 0 0 8 3.5c-2.798 0-5.047 1.439-6.819 4.432.842 1.465 1.792 2.568 2.855 3.323zm2.948-1.532a2 2 0 0 0 2.74-2.738l-2.74 2.738zm-.707-.707l2.74-2.738a2 2 0 0 0-2.74 2.738z" />
            </svg>
          </Tooltip>
          <Tooltip content="Hide Password" direction="top" type="icon">
            <svg
              className="hcl-icon-visibility-on"
              aria-label="Hide password"
              width="16"
              height="16"
              viewBox="0 0 16 11"
              style={
                !isIconVisible ? { display: 'none' } : { display: 'initial' }
              }
            >
              <path d="M8 7.5c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 1c-1.7 0-3-1.3-3-3s1.3-3 3-3 3 1.3 3 3-1.3 3-3 3z" />
              <path d="M8 10c2.8 0 5.1-1.5 6.9-4.6C13.1 2.5 10.8 1 8 1 5.2 1 3 2.4 1.2 5.4 2.9 8.6 5.2 10 8 10zM8 0c3.3 0 6 1.8 8.1 5.4C14 9.2 11.3 11 8 11S2 9.2 0 5.5C2 1.9 4.6 0 8 0z" />
            </svg>
          </Tooltip>
        </>
      )}
    </button>
  );

  useEffect(() => {
    if (restProps.value || restProps.value === '') {
      setValue(restProps.value);
    }
  }, [restProps.value]);

  useEffect(() => {
    if (isIconVisible) {
      setType('text');
    } else {
      setType('password');
    }
  }, [isIconVisible]);

  return (
    <>
      <input
        ref={inputRef}
        className={classnames}
        type={typechange}
        disabled={disabled}
        {...restProps}
        value={value}
        onChange={event => {
          setValue(event.currentTarget.value);
          restProps.onChange(event);
        }}
      />
      {passwordIcon}
    </>
  );
};

Password.propTypes = {
  /** Additional class name to be given to <input> tag. */
  className: PropTypes.string,
  /** Specifying the <input> tag is disabled or not. */
  disabled: PropTypes.bool,
  /** Unique Id for the <input> tag. */
  id: PropTypes.string,
  /** Placeholder text for the <input> tag. */
  placeholder: PropTypes.string,
  /** Value of the input field. */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** The default Value of the field on rendering. */
  defaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** Event to subscribe when the value of the Input field changes. */
  onChange: PropTypes.func,
  /** Event to subscribe when the Input field is clicked. */
  onClick: PropTypes.func,
  /** Event to subscribe when the Input field is focused. */
  onFocus: PropTypes.func,
  /** Event to subscribe when the Input field is Blurred. */
  onBlur: PropTypes.func
};

Password.defaultProps = {
  className: '',
  disabled: false,
  onChange: () => {},
  onClick: () => {},
  onFocus: () => {},
  onBlur: () => {}
};

export default Password;
