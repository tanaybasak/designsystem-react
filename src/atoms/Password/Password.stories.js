import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { text, boolean } from '@storybook/addon-knobs';
//@update-path-build-start
import Password from './Password';
//@update-path-build-end

storiesOf('Components/Password', module)
  .add(
    'default',
    () => (
      <div className="hcl-form-group">
        <Password
          className=""
          disabled={boolean('Disabled', false)}
          onBlur={action('onBlur')}
          onChange={action('onChange')}
          onClick={action('onClick')}
          onFocus={action('onFocus')}
          placeholder={text('Placeholder', 'Placeholder Text')}
          aria-label="text input"
        />
      </div>
    ),
    {
      info: {
        text: `Description About Password Component`,
        document: ['Password'],
        className: 'hcl-col-12 hcl-col-sm-8 hcl-col-lg-4'
      }
    }
  )
  .add(
    'with error',
    () => (
      <div className="hcl-form-group">
        <Password
          className=""
          disabled={boolean('Disabled', false)}
          onBlur={action('onBlur')}
          onChange={action('onChange')}
          onClick={action('onClick')}
          onFocus={action('onFocus')}
          placeholder={text('Placeholder', 'Placeholder Text')}
          aria-label="text input"
          data-invalid="true"
        />
        <label className="hcl-label" htmlFor="labeltext">
          Label text
        </label>
        <div className="hcl-error-msg">Validation message here</div>
      </div>
    ),
    {
      info: {
        text: `Description About Password Component`,
        document: ['Password'],
        className: 'hcl-col-12 hcl-col-sm-8 hcl-col-lg-4'
      }
    }
  )
  .add(
    'with helpertext',
    () => (
      <div className="hcl-form-group">
        <Password
          className=""
          disabled={boolean('Disabled', false)}
          onBlur={action('onBlur')}
          onChange={action('onChange')}
          onClick={action('onClick')}
          onFocus={action('onFocus')}
          placeholder={text('Placeholder', 'Placeholder Text')}
          aria-label="text input"
        />
        <label className="hcl-label" htmlFor="labeltext">
          Label text
        </label>
        <div className="hcl-helper-text">Optional Helper text goes here</div>
      </div>
    ),
    {
      info: {
        text: `Description About Password Component`,
        document: ['Password'],
        className: 'hcl-col-12 hcl-col-sm-8 hcl-col-lg-4'
      }
    }
  );
