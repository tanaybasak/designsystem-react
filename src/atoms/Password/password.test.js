import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Password from './Password';

describe('<Password> component', () => {
  it('snapshots/renders Password correctly', () => {
    const tree = renderer.create(<Password />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('simulate change event in Password and is called Once', () => {
    const mockCallBack = jest.fn();
    const wrapper = shallow(<Password onChange={mockCallBack} />);
    wrapper
      .find('input[type="password"]')
      .simulate('change', { currentTarget: { value: 'Hello' } });
    expect(mockCallBack.mock.calls.length).toEqual(1);
    expect(wrapper.find('input[type="password"]').props().value).toBe('Hello');
  });

  it('simulate click, onfocus event in Password and is called Once', () => {
    const mockCallBackclick = jest.fn().mockReturnValue(undefined);
    const mockCallBackfocus = jest.fn().mockReturnValue(undefined);
    const mockCallBackchange = jest.fn().mockReturnValue(undefined);
    const mockCallBackblur = jest.fn().mockReturnValue(undefined);
    const wrapper = shallow(<Password />);
    wrapper.find('input[type="password"]').simulate('click', { target: {} });
    wrapper.find('input[type="password"]').simulate('focus', { target: {} });
    wrapper
      .find('input[type="password"]')
      .simulate('change', { currentTarget: { value: 'Hello' } });
    wrapper.find('input[type="password"]').simulate('blur', { target: {} });
    expect(mockCallBackclick()).toEqual(undefined);
    expect(mockCallBackfocus()).toEqual(undefined);
    expect(mockCallBackchange()).toEqual(undefined);
    expect(mockCallBackblur()).toEqual(undefined);
  });
});
