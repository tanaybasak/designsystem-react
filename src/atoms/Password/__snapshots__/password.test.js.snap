// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Password> component snapshots/renders Password correctly 1`] = `
Array [
  <input
    className="hcl-form-control"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    type="password"
    value=""
  />,
  <button
    aria-label="Show password"
    className="hcl-password-visibility"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      className="hcl-icon-visibility-off"
      height="16"
      onBlur={[Function]}
      onClick={null}
      onFocus={[Function]}
      onKeyPress={null}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "display": "initial",
        }
      }
      tabIndex="0"
      viewBox="0 0 16 16"
      width="16"
    >
      <path
        d="M11.846 3.45L15.293.007 16 .714l-3.284 3.281c1.261.902 2.377 2.212 3.347 3.93C14.02 11.642 11.333 13.5 8 13.5c-1.392 0-2.667-.324-3.822-.973L.703 16l-.706-.708 3.323-3.32C2.071 11.042.976 9.694.035 7.924 2.012 4.308 4.667 2.5 8 2.5c1.395 0 2.677.317 3.846.95zm-6.928 8.338c.944.477 1.97.712 3.082.712 2.795 0 5.076-1.483 6.907-4.568-.866-1.417-1.833-2.486-2.91-3.219l-1.55 1.55a3 3 0 0 1-4.185 4.182l-1.344 1.343zm-.882-.533l1.518-1.517A3 3 0 0 1 9.74 5.556l1.364-1.363A7.02 7.02 0 0 0 8 3.5c-2.798 0-5.047 1.439-6.819 4.432.842 1.465 1.792 2.568 2.855 3.323zm2.948-1.532a2 2 0 0 0 2.74-2.738l-2.74 2.738zm-.707-.707l2.74-2.738a2 2 0 0 0-2.74 2.738z"
      />
    </svg>
    <svg
      aria-label="Hide password"
      className="hcl-icon-visibility-on"
      height="16"
      onBlur={[Function]}
      onClick={null}
      onFocus={[Function]}
      onKeyPress={null}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      tabIndex="0"
      viewBox="0 0 16 11"
      width="16"
    >
      <path
        d="M8 7.5c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 1c-1.7 0-3-1.3-3-3s1.3-3 3-3 3 1.3 3 3-1.3 3-3 3z"
      />
      <path
        d="M8 10c2.8 0 5.1-1.5 6.9-4.6C13.1 2.5 10.8 1 8 1 5.2 1 3 2.4 1.2 5.4 2.9 8.6 5.2 10 8 10zM8 0c3.3 0 6 1.8 8.1 5.4C14 9.2 11.3 11 8 11S2 9.2 0 5.5C2 1.9 4.6 0 8 0z"
      />
    </svg>
  </button>,
]
`;
